core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -267,376 +302,264 @@ protected void doRun() throws Exception {
-                    } catch (Throwable e) {
-                        failAndRemoveShard(shardRouting, indexService, true, "failed updating shard routing entry", e);
-                    }
+            AllocatedIndex<? extends Shard> indexService = indicesService.indexService(shardId.getIndex());
+            assert indexService != null;
+            if (indexService != null) {
+                Shard shard = indexService.getShardOrNull(shardId.id());
+                if (shard == null) {
+                    createShard(nodes, routingTable, shardRouting, indexService);
+                } else {
+                    updateShard(event, nodes, shardRouting, shard);
                 }
             }
-
-            if (shardRouting.initializing()) {
-                applyInitializingShard(event.state(), indexService, shardRouting);
-            }
         }
     }
 
-    private void applyInitializingShard(final ClusterState state, IndexService indexService, final ShardRouting shardRouting) {
-        final RoutingTable routingTable = state.routingTable();
-        final DiscoveryNodes nodes = state.getNodes();
-        final int shardId = shardRouting.id();
-
-        if (indexService.hasShard(shardId)) {
-            IndexShard indexShard = indexService.getShard(shardId);
-            if (indexShard.state() == IndexShardState.STARTED || indexShard.state() == IndexShardState.POST_RECOVERY) {
-                // the master thinks we are initializing, but we are already started or on POST_RECOVERY and waiting
-                // for master to confirm a shard started message (either master failover, or a cluster event before
-                // we managed to tell the master we started), mark us as started
-                if (logger.isTraceEnabled()) {
-                    logger.trace("{} master marked shard as initializing, but shard has state [{}], resending shard started to {}",
-                        indexShard.shardId(), indexShard.state(), nodes.getMasterNode());
-                }
-                if (nodes.getMasterNode() != null) {
-                    shardStateAction.shardStarted(shardRouting,
-                        "master " + nodes.getMasterNode() + " marked shard as initializing, but shard state is [" + indexShard.state() + "], mark shard as started",
-                        SHARD_STATE_ACTION_LISTENER);
-                }
-                return;
-            } else {
-                if (indexShard.ignoreRecoveryAttempt()) {
-                    logger.trace("ignoring recovery instruction for an existing shard {} (shard state: [{}])", indexShard.shardId(), indexShard.state());
+    private void createShard(DiscoveryNodes nodes, RoutingTable routingTable, ShardRouting shardRouting,
+                             AllocatedIndex<? extends Shard> indexService) {
+        assert shardRouting.initializing();
+
+        if (shardRouting.initializing()) {
```

Comment:
we do we need this if? it was started we should have failed it in the beginning?